spec/gears/color_spec.lua

```diff
@@ -291,6 +292,233 @@ describe("gears.color", function()
+            it(col[1], function()
+                hsla_match(col[2], {color.rgba_to_hsla(col[1])})
+            end)
+        end
+    end)
+
+    -- This section is sort of patchy as HSLA to RGBA isn't an exact number
+    -- because of rounding and floating point inprecision as a whole. I've
+    -- adjusted the random colors to be able to work back and forth, but a
+    -- few of a single RGB channel had to be adjusted by 1 or 2.
+    --
+    -- This covers hsl_to_rgb since it calls hsla_to_rgba.
+    describe("hsla_to_rgba", function()
+        local function rgba_match(exp, pass)
+            -- we're going to match the strings here since stringify should be
+            -- working as tested above
+            assert.is.same(exp, color.stringify(pass, false, 8))
+        end
+        for _,col in ipairs(colors) do
+            it(col[1], function()
+                local h,s,l = col[2][1], col[2][2] / 100, col[2][3] / 100
+                rgba_match(col[1] .. "ff", {color.hsla_to_rgba({h,s,l,1})})
+            end)
+        end
+    end)
+
+    describe("lighten 25%", function()
+        for _,col in ipairs(colors) do
+            it(col[1], function()
+                assert.is.same(col[3] .. "ff", color.stringify({color.lighten(col[1], 0.25)}, false, 8))
+            end)
+        end
+    end)
+
+    describe("darken 25%", function()
+        for _,col in ipairs(colors) do
+            it(col[1], function()
+                assert.is.same(col[4] .. "ff", color.stringify({color.darken(col[1], 0.25)}, false, 8))
+            end)
+        end
+    end)
+
+    describe("invert", function()
+        for _,col in ipairs(colors) do
+            it(col[1], function()
+                assert.is.same(col[5] .. "ff", color.stringify({color.invert(col[1])}, true, 8))
+            end)
+        end
+    end)
+
+    -- 0 of first, 100 of second
+    describe("mix 0%", function()
+        for _,col in ipairs(colors) do
+            it(col[1], function()
+                assert.is.same(col[5] .. "ff", color.stringify({color.mix(col[1], col[5], 0)}, true, 8))
+            end)
+        end
+    end)
+
+    -- 1/4 of first, 3/4 of second
+    describe("mix 25%", function()
+        for _,col in ipairs(colors) do
+            it(col[1], function()
+                assert.is.same(col[6] .. "ff", color.stringify({color.mix(col[1], col[5], 0.25)}, true, 8))
+            end)
+        end
+    end)
+
+    -- 3/4 of first, 1/4 of second
+    describe("mix 75%", function()
+        for _,col in ipairs(colors) do
+            it(col[1], function()
+                assert.is.same(col[7] .. "ff", color.stringify({color.mix(col[1], col[5], 0.75)}, true, 8))
+            end)
+        end
+    end)
+
+    -- 100 of first, 0 of second
+    describe("mix 100%", function()
+        for _,col in ipairs(colors) do
+            it(col[1], function()
+                assert.is.same(col[1] .. "ff", color.stringify({color.mix(col[1], col[5], 1)}, true, 8))
```

Comment:
does this repetition of stringify every time looks fine for you? i would suggest having `color.hex_mix` , `hex_darker` and other similar proxies or find some other solution to have more usable output from those new functions

Reply:
That's not a bad idea.

I was actually thinking about creating a complete `Color` object which we could use similar to how SCSS is put together. I noticed in the GTK helper that the color was accessible through `gtk_color.red`, `.green`, `.blue`, and `.alpha`. That is my idea for a good object including the additional HSL code. Ideally I wanted to be able to do something like `color:lighten(0.25)` and do all the magic behind the scenes, afterwards doing something like `color:str()` to get the stringify'd output.

Reply:
i like the idea with object, @Elv13 do you have any opinion on this?

Reply:
So, similar to `gears.matrix`, I guess.

Reply:
btw i think `:hex()` would be better than `:str()` since it's more explicit

Reply:
> btw i think `:hex()` would be better than `:str()` since it's more explicit

For me, I'd expect `:str()` to return a string representation of the color in `#rrggbb[aa]` format, and `:hex()` to return a decimal (that's actually how I have it right now). But I'll leave that for the group to decide since it's just a function name change. :)

I have a working Color object right now that I've played with based upon how `gears.matrix` is built. I'm looking at making it `gears.color.Color`. @psychon @actionless I'll open a different PR for this direction when I get it fixed up.

Reply:
> my point what there are several possible string representations for color (let's say `rgba(1, 2, 3, 4)`) so i was thinking from position what it will be the most understandable which exactly text representation is used

We can discuss how to name the different output representations in the other PR once I get it created.

> > :hex() to return a decimal
> 
> that sounds not logical to me: decimal is base of 10, hex is base of 16

When you print a hexadecimal number in lua do you see "0xf" or "15"? ;) Just pulling your leg.
But I know what you mean. They'll just be function names we can rename, so it's all good.

Reply:
so your whole idea to now create functions and in new PR create an object which will be using those functions to implement color operations on it?

Reply:
Ideally everything will be done within the `Color` object itself.

I sort of have it set up like this right now (we can rename output functions later):
```lua
Color = (require 'gears.color').Color
c = Color.new("#ffcc99")
print(c:as_str()) -- prints "#ffcc99"
print(c:as_hex()) -- prints "4291598847"
print(c.red) -- prints 1
print(c.green) -- prints 0.8
print(c.blue) -- prints 0.6
print(c.hue) -- prints 30.0
print(c.saturation) -- prints 1.0
print(c.lightness) -- prints 0.8
print(c.opacity) -- prints 1
c2 = Color.new("#99ccff")
print(c1:mix(c2)) -- defaults to 0.5 mix, half and half; prints #ccccccff
print(c1:mix(c2, 0)) -- prints #ffcc99ff
print(c1:mix(c2, 1)) -- prints #99ccffff
```

You can do a function on the object itself and it will return a new color (unless we want to make something like `c1:mix(c2)` destroy the old color). So we'd have `color:lighten`, `color:darken`, etc.

The other PR I'll make (unless I just wipe this one and rename it) will essentially replace this one if we want to go that route.

Reply:
i would suggest implementing object right in this PR

especially because release is soon and if we merge one API and next will replace it with another API that will mean deprecating old API and providing backward compatibility

Reply:
and regarding naming, you could just google "hex colors" and observe what it's common name of "#0088ff" representation

Reply:
other than points above : i really like color object idea and appreciate the time you investing into making it happen :+1:

Reply:
Alright! Sounds good! I'll do it here. :)